src/v/pandaproxy/server.h

```diff
@@ -108,6 +106,23 @@ template<typename service_t>
 class ctx_server : public server {
 public:
     using server::server;
+
+    struct context_t : server::context_t {
+        service_t& service;
+    };
+
+    // request_t restores the type of the context passed in.
+    struct request_t final : server::request_t {
+        // Implicit constructor from type-erased server::request_t.
+        request_t(server::request_t&& impl) // NOLINT
+          : server::request_t(std::move(impl)) {}
+        // Type-restored context
```

Comment:
this looks correct. I haven't seen this type of design pattern before. Is there a name for it? specifically in the context of how the code has been reorganized to make the generic server?

Reply:
Not that I know of. I had two criteria:

* Avoid making the base a template
* Avoid heap allocation

References was the obvious choice after making `context_t` usage cleaner. I'm open to ideas, the type-erase and type-restore are a little further from each other then desired.

Reply:
seems fine to me thx.